### DISCUSSION TITLE
How to load only Contacts modified after a specific timestamp?

### DISCUSSION THREAD

**Discussion by M00r14rty** [vestrel00/contacts-android#199] [Q&A] [answered]
Hi everyone,
First of all, this library rocks!

Now, my problem. I would like to load only Contacts modified after a specific timestamp. Unfortunately, the `Fields.Contact.LastUpdatedTimestamp` is updated every sync or after any interaction with the contact, making this field completely unuseful.

I think I can check the `ContactsContract.RawContacts.VERSION` or `ContactsContract.DataColumns.DATA_VERSION`. How can I retrieve this data using this library?
Is there a better way to solve my problem?

Thanks!

**Comment by vestrel00** [comment, marked as answer]
Hi @M00r14rty, this is a good question!

> How to load only Contacts modified after a specific timestamp?

As you pointed out in your question, I believe you are already doing something like this,

```kotlin
fun loadContactsModifiedAfter(timestampMillis: Long): List<Contact> = Contacts(this)
    .query()
    .where { Contact.LastUpdatedTimestamp greaterThan timestampMillis }
    .find()
```

This is technically the answer to your question. But I can see it's not working the way you want it to, so let's dig a little deeper 🙂

> First of all, this library rocks!

Thanks! ❤️❤️❤️❤️❤️❤️

> load only Contacts modified after a specific timestamp

The `Fields.Contact.LastUpdatedTimestamp` should be what you need to use. The [official documentation](https://developer.android.com/reference/android/provider/ContactsContract.ContactsColumns#CONTACT_LAST_UPDATED_TIMESTAMP) for it is,

<img width="869" alt="Screen Shot 2022-03-22 at 4 33 05 AM" src="https://user-images.githubusercontent.com/4332753/159506538-cef420ce-ca55-481e-873c-e729aa94e1ca.png">

If the official documentation is correct, and from my observations, this timestamp should only be updated when any one of the following events occur;

- Any row in the Data table (e.g. email, phone) belonging to the contact is updated, inserted, or deleted.
- Any row in the RawContacts table belonging to the contact is updated, inserted, or deleted.
- Any row in the Contacts table is updated or inserted.

> updated every sync

Any of the above events could occur when a sync is performed. It could be due to the Contacts Provider performing some aggregation or 3rd party apps' SyncAdapters writing to the RawContacts or Data table rows. Unfortunately, we have no control over this.

> after any interaction with the contact

I think this is a good callout! If you look at the code in `ContactsContract`, there are columns in the Contacts table that may get automatically updated when;

- The contact is contacted (e.g. messaged via SMS or called via telephony).
- The contact presence changes (e.g. during chat goes online/offline).

> I think I can check the `ContactsContract.RawContacts.VERSION` or `ContactsContract.DataColumns.DATA_VERSION`.

These rows should be unaffected "after any interaction with the contact". However, they may be "updated every sync".  Both of these are also just integers that start at 0 or 1. They are not timestamps, so you cannot use these to "load only Contacts modified after a specific timestamp". Though, I suppose that you can save the `RawContacts.VERSION` in your app using a database table or shared preference and associate a timestamp to it but that involves doing things outside the Contacts Provider and it could get messy.

> How can I retrieve this data using this library?

These fields are currently not in the library. We would have to add them in the code first.  However, I don't think these will solve your problem.

> Is there a better way to solve my problem?

We could add custom data that we can put a timestamp into. Then, in your app, you can set the timestamp whenever your app "modifies" any data belonging to a RawContact or just whenever you want to set the timestamp. This timestamp will be unaffected by the Contacts Provider's  "sync" or "interaction with the contact". 

There are two limitations with this approach;

- This custom data timestamp can only be modified by your app (or other apps using the custom data). 
    - This means that the timestamp will not be updated when modifying the contact using other contacts apps such as the builtin AOSP Contacts app or the [Google Contacts app](https://play.google.com/store/apps/details?id=com.google.android.contacts). 
- This custom data will also not be synced across devices unless you implement a sync adapter for it in your app.

If you are okay with these limitations, then [please create an issue](https://github.com/vestrel00/contacts-android/issues/new). I might call the module `customdata-lastupdatedtimestamp`.

Let me know!

## Other considerations

If you search for "time" in the `ContactsContract` file, you will find that `CONTACT_LAST_UPDATED_TIMESTAMP` is the only related field that is related to your problem, unfortunately.

Also, please let me know if the builtin AOSP Contacts app or the [Google Contacts app](https://play.google.com/store/apps/details?id=com.google.android.contacts) has features that allow you to do exactly what you want to do (and how you want it to behave). This will give us a hint if it is truly something we can do with the Contacts Provider. I can guarantee you that this library will support any and all features in the AOSP Contacts app or the [Google Contacts app](https://play.google.com/store/apps/details?id=com.google.android.contacts) and more.

## FYI

I changed the title of this discussion to "How to load only Contacts modified after a specific timestamp?" because it is the underlying question. Unfortunately, I don't think `RawContacts.VERSION` or `DataColumns.DATA_VERSION` is the answer 😔

**Comment by M00r14rty** [comment]
Thank you @vestrel00 for your exhaustive answer.

As you pointed out, I already tried to use the where clause in the query.

Unfortunately, I didn't find any useful field in the AOSP Contacts, that's why I decided to test a different approach. My idea is to abandone the timestamp and retrieve the versions of all contact rows to store them as unique hash value.
Every time my app starts, it retrieve all the versions and calculate the hash to compare it with the stored one. If they are different, then I can reload all the contacts. That's just an idea, I suppose it is faster than load and manipulate them every time. I need to test it, if you are interested I will post the results.

**Reply by vestrel00** [reply]
I'm interested in the results! Please let me know how it goes. If your approach works, please file an issue for the fields you'd like to be added to this library 😁
